### DISCUSSION TITLE
pip-compile manager not working as expected (with minimal repro)

### DISCUSSION THREAD

**Discussion by jab** [renovatebot/renovate#12432] [Request Help]
I recently discovered Renovate and was super psyched to try it with one of my open source Python projects (https://github.com/jab/bidict) for which I currently run `pip-compile` locally to upgrade dependencies. I followed the https://docs.renovatebot.com/modules/manager/pip-compile/ docs (and also found [this other related discussion](https://github.com/renovatebot/renovate/discussions/10407) here) but haven't yet gotten Renovate to work as expected.

To help track this down, I created a minimal reproduction here: https://github.com/jab/renovate-pip-compile-repro

This repo has a single `requirements.in` file (namely `requirements/docs.in`), with a single dependency in it (namely `Sphinx`).

When I run `pip-compile -U requirements/docs.in && git diff` in this repo locally, I get the expected result for the one (transitive) dependency pinned in the docs.txt lockfile that needs upgrading:
```diff
diff --git a/requirements/docs.txt b/requirements/docs.txt
index edaed00..7f45bca 100644
--- a/requirements/docs.txt
+++ b/requirements/docs.txt
@@ -22,7 +22,7 @@ jinja2==3.0.2
     # via sphinx
 markupsafe==2.0.1
     # via jinja2
-packaging==21.0
+packaging==21.2
     # via sphinx
 pygments==2.10.0
     # via sphinx
```
But when Renovate bot runs against this repo, it fails to raise any PR for this available upgrade, even though [the debug output](https://app.renovatebot.com/dashboard#github/jab/renovate-pip-compile-repro/497139494) shows that is finding the right `requirements/docs.in` input file:
> DEBUG: Matched 1 file(s) for manager pip-compile: requirements/docs.in

Please let me know how I can get Renovate working as expected with this minimal repro repo, and then hopefully I can get it to work with my real repo. Thanks!

**Comment by rarkins** [comment]
There's a very long discussion here: #2334

**Reply by jab** [reply]
Thanks @rarkins, I can read that, but is there a tl;dr by any chance? It seems to me based on my minimal repro that the https://docs.renovatebot.com/modules/manager/pip-compile/ docs are not working as advertised. Can you just say real quick whether that's the case? Thank you!

**Reply by jab** [reply]
(I know that those docs do advertise "limited functionality" e.g. the "Assumption of .in/.txt" but that seems just fine for the simple use case I put in my repro!)

**Reply by rarkins** [reply]
What does your `packageFiles with updates` log message indicate?

**Reply by jab** [reply]
Quoting https://app.renovatebot.com/dashboard#github/jab/renovate-pip-compile-repro/497139494
```
DEBUG: packageFiles with updates
{
  "config": {}
}
```
Thanks for your continued help with this!

**Reply by rarkins** [reply]
I took a look and found two problems:
1. Currently dependencies are ignored unless they have a constraint or version
2. We weren't passing the file name to `pip-compile` so it was assuming `requirements.in` or `setup.py`

You can see the result in https://github.com/renovate-reproductions/renovate-pip-compile-repro/pull/1 once I changed those two things. See #12433 for the filename fix

BTW I don't have time right now to add python version detection (e.g. run with 3.9 instead of 3.10) but you're welcome to submit a feature request if you think it's important

**Reply by viceice** [reply]
Should we pin back python 3.9 as latest docker image until most python issues are fixed?

**Reply by jab** [reply]
Glad that my repro here helped you identify the filename bug, and great you have a fix submitted already, nice! (Since I’m new to Renovate, once that fix is merged, about how long before it makes it to production?)

Problem number 1 — that dependencies specified in `*.in` files are ignored if they have no constraint or version — seems more serious though. I think that bears a conspicuous mention in your docs for the pip-compile manager. If you look at how many (if not most) Python projects are using those `*.in` files, they’re not including any pin or constraint there, since they want the latest version available without bound (and rely on generated lockfiles to include the actual version pins along with automated testing to catch any problems before merging).

**Reply by jab** [reply]
(Oh, and the Python 3.9 vs. 3.10 issue isn’t important at all in my case, and I think same goes for most other pip-compile users)

**Reply by jab** [reply]
Is there an issue I can subscribe to that tracks fixing problem number 1 above? That’s a blocker for me in adopting Renovate. I guess I could resort to a hacky workaround (like adding an artificial constraint like `>=0` for each unconstrained dependency), but that could be brittle and I’m not sure I want to do that.

(Btw, I tested that workaround for problem number 1 in my "minimal repro" repo and verified that I then hit problem number 2 (the filename bug) as expected: https://app.renovatebot.com/dashboard#github/jab/renovate-pip-compile-repro/497199480
<details>
<summary>click to reveal log output corresponding to the bug</summary>

```
DEBUG: Executing command(branch="renovate/pip-compile-refresh")
{
  "command": [
    "docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\""
  ]
}
DEBUG: rawExec err(branch="renovate/pip-compile-refresh")
{
  "err": {
    "killed": false,
    "code": 2,
    "signal": null,
    "cmd": "docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\"",
    "stdout": "Collecting pip-tools\n  Downloading pip_tools-6.4.0-py3-none-any.whl (47 kB)\nCollecting click>=7\n  Downloading click-8.0.3-py3-none-any.whl (97 kB)\nRequirement already satisfied: setuptools in /usr/local/python/3.10.0/lib/python3.10/site-packages (from pip-tools) (57.4.0)\nCollecting pep517\n  Downloading pep517-0.12.0-py2.py3-none-any.whl (19 kB)\nCollecting wheel\n  Downloading wheel-0.37.0-py2.py3-none-any.whl (35 kB)\nRequirement already satisfied: pip>=21.2 in /usr/local/python/3.10.0/lib/python3.10/site-packages (from pip-tools) (21.3.1)\nCollecting tomli>=1.1.0\n  Downloading tomli-1.2.2-py3-none-any.whl (12 kB)\nInstalling collected packages: tomli, wheel, pep517, click, pip-tools\nSuccessfully installed click-8.0.3 pep517-0.12.0 pip-tools-6.4.0 tomli-1.2.2 wheel-0.37.0\n",
    "stderr": "Usage: pip-compile [OPTIONS] [SRC_FILES]...\nTry 'pip-compile -h' for help.\n\nError: Invalid value: If you do not specify an input file, the default is requirements.in or setup.py\n",
    "message": "Command failed: docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\"\nUsage: pip-compile [OPTIONS] [SRC_FILES]...\nTry 'pip-compile -h' for help.\n\nError: Invalid value: If you do not specify an input file, the default is requirements.in or setup.py\n",
    "stack": "Error: Command failed: docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\"\nUsage: pip-compile [OPTIONS] [SRC_FILES]...\nTry 'pip-compile -h' for help.\n\nError: Invalid value: If you do not specify an input file, the default is requirements.in or setup.py\n\n    at ChildProcess.exithandler (child_process.js:390:12)\n    at ChildProcess.emit (events.js:400:28)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
  }
}
DEBUG: Failed to pip-compile(branch="renovate/pip-compile-refresh")
{
  "err": {
    "killed": false,
    "code": 2,
    "signal": null,
    "cmd": "docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\"",
    "stdout": "Collecting pip-tools\n  Downloading pip_tools-6.4.0-py3-none-any.whl (47 kB)\nCollecting click>=7\n  Downloading click-8.0.3-py3-none-any.whl (97 kB)\nRequirement already satisfied: setuptools in /usr/local/python/3.10.0/lib/python3.10/site-packages (from pip-tools) (57.4.0)\nCollecting pep517\n  Downloading pep517-0.12.0-py2.py3-none-any.whl (19 kB)\nCollecting wheel\n  Downloading wheel-0.37.0-py2.py3-none-any.whl (35 kB)\nRequirement already satisfied: pip>=21.2 in /usr/local/python/3.10.0/lib/python3.10/site-packages (from pip-tools) (21.3.1)\nCollecting tomli>=1.1.0\n  Downloading tomli-1.2.2-py3-none-any.whl (12 kB)\nInstalling collected packages: tomli, wheel, pep517, click, pip-tools\nSuccessfully installed click-8.0.3 pep517-0.12.0 pip-tools-6.4.0 tomli-1.2.2 wheel-0.37.0\n",
    "stderr": "Usage: pip-compile [OPTIONS] [SRC_FILES]...\nTry 'pip-compile -h' for help.\n\nError: Invalid value: If you do not specify an input file, the default is requirements.in or setup.py\n",
    "message": "Command failed: docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\"\nUsage: pip-compile [OPTIONS] [SRC_FILES]...\nTry 'pip-compile -h' for help.\n\nError: Invalid value: If you do not specify an input file, the default is requirements.in or setup.py\n",
    "stack": "Error: Command failed: docker run --rm --name=renovate_python --label=renovate_child -v \"/mnt/renovate/gh/jab/renovate-pip-compile-repro\":\"/mnt/renovate/gh/jab/renovate-pip-compile-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -w \"/mnt/renovate/gh/jab/renovate-pip-compile-repro/requirements\" docker.io/renovate/python bash -l -c \"pip install --user pip-tools && pip-compile\"\nUsage: pip-compile [OPTIONS] [SRC_FILES]...\nTry 'pip-compile -h' for help.\n\nError: Invalid value: If you do not specify an input file, the default is requirements.in or setup.py\n\n    at ChildProcess.exithandler (child_process.js:390:12)\n    at ChildProcess.emit (events.js:400:28)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
  }
}
```
</details>
)

**Reply by rarkins** [reply]
Coming from a JS background, I think it's a terrible convention to not at least pin dependencies to a major version. When one of your dependencies bumps a major version to signify a breaking change.. you just break?

**Reply by rarkins** [reply]
(I know that the tree is "frozen" by pip-compile, but the next time you run it, you'll get the breaking change and have to figure out why)

**Reply by jab** [reply]
https://alexgaynor.net/2021/oct/07/whats-in-a-version-number/

**Reply by rarkins** [reply]
See https://github.com/renovatebot/renovate/issues/6955

**Reply by jab** [reply]
I used to think that way too (see my own https://github.com/python-trio/trio/issues/1182!), but I've since been convinced that team "don't blindly trust semver it's much more nuanced than that" is correct.

**Reply by jab** [reply]
I'm sorry to hear that, I'm not sure why you think it's pointless and I only meant to be helpful (as I already have been elsewhere in this thread). Maybe I should be done with giving Renovate a chance then too. What a disappointing first impression.

**Reply by rarkins** [reply]
Sorry, but I'm not a fan of the "Software versioning is hard so why even try?" justification for YOLO dependency management, and thought you considered it to be all you needed to say on the subject. As you followed up with a further comment on the topic and not just a URL without comment, I have deleted my earlier response.

Particularly with the advent of automated tools which can alert you of new versions, I don't think there's any justification for unbound dependencies, even if some people find it hard. The future of software is reliable, deterministic dependency management and not "yeah so what if it breaks every now and then, that's life". Apps should use pinned/locked dependencies, while libraries which use unbound ranges and break users will lose appeal versus those who spend the time to maintain stable ranges for their downstream users.

**Reply by jab** [reply]
Thanks for engaging more productively here. Given your very first response when I opened this discussion, which could also be interpreted as dismissive, but which I tried to interpret with better intentions than that, I'm sorry you didn't do the same for me.

More important, the project I'm trying to use Renovate with doesn't even have any runtime dependencies. All the unbounded dependencies which I'm trying to have a tool like Renovate manage for me **are development, documentation, or testing dependencies**.

So other people's code that depends on my project **is not affected by the fact that my dev/docs/test-time dependencies are unbounded**, since they don't inherit any of those dependencies transitively.

I prefer to keep my dev/docs/test-time dependencies unbounded because it allows me to get more out of a tool like Renovate when combined with the required PR status checks I have set up, which catch any major upgrades that do turn out to break my docs, tests, linters, etc. (which in the 11+ years of maintaining this project has actually happened quite rarely), and when breakage does occur, I either do go add a constraint to the corresponding `docs.in` file, or I go change whatever it was in my docs that didn't work with the latest major version of that dependency to work with it.

At no time would this ever introduce breakage to my main branch, let alone to my users.

I think this is a sound use of unbounded dependencies, and so any philosophical arguments (which don't even apply in this case) should not delay addressing https://github.com/renovatebot/renovate/pull/12433 or https://github.com/renovatebot/renovate/issues/6955, both of which are needed before Renovate can be used by projects like mine.

**Reply by viceice** [reply]
We, the renovate maintainer, need to decide what we priorize. That are at first bugs. Then we implement features we need. Then we review community contributions.

So we have a lot of work, so features like #6955  will need a community contribution to get it done. 🤷‍♂️

**Reply by jab** [reply]
Understood, I'm a maintainer too. Hope it's helpful for your prioritization to know that the current level of pip-compile support makes it unusable for projects like mine (which also includes the Pallets projects I help maintain, see e.g. [the `*.in` files that Flask uses](https://github.com/pallets/flask/tree/main/requirements), which I was also interested to try to use Renovate with, and which could in turn help Renovate attract more users given Flask's popularity). And again, please mention this limitation on the https://docs.renovatebot.com/modules/manager/pip-compile/ docs, it could save future users like me significant time and frustration.

**Reply by rarkins** [reply]
Source: https://github.com/renovatebot/renovate/blob/main/lib/manager/pip-compile/readme.md
